modules/javafx.controls/src/main/java/javafx/scene/control/input/InputMap2.java

```diff
@@ -0,0 +1,546 @@
+                en.behavior = null;
+                en.behaviorValue = null;
+            }
+        }
+    }
+
+    <T extends Event> void map(IBehavior behavior, EventType<T> type, boolean consume, boolean tail, EventHandler<T> handler) {
+        if (consume) {
+            extendHandlers(behavior, type, tail, new EventHandler<T>() {
+                @Override
+                public void handle(T ev) {
+                    handler.handle(ev);
+                    ev.consume();
+                }
+            });
+        } else {
+            extendHandlers(behavior, type, tail, handler);
+        }
+    }
+    
+    <T extends Event> void map(IBehavior behavior, EventCriteria<T> criteria, boolean consume, boolean tail, EventHandler<T> handler) {
+        EventType<T> type = criteria.getEventType();
+        extendHandlers(behavior, type, tail, new EventHandler<T>() {
+            @Override
+            public void handle(T ev) {
+                if (criteria.isEventAcceptable(ev)) {
+                    handler.handle(ev);
+                    if (consume) {
+                        ev.consume();
+                    }
+                }
+            }
+        });
+    }
+
+    private <T extends Event> void extendHandlers(IBehavior behavior, EventType<T> t, boolean tail, EventHandler<T> h) {
+        Objects.nonNull(behavior);
+        Entry en = addListenerIfNeeded(t);
+
+        HList handlers = HList.from(en.behaviorValue);
+        handlers.add(h, tail);
+        en.behavior = behavior;
+        en.behaviorValue = handlers;
+    }
+    
+    /**
+     * Adds a user-specified function under the given function tag.
+     * This function will override any function set by the behavior.
+     * @param tag the function tag
+     * @param function the function
+     */
+    public void func(FunctionTag tag, Runnable function) {
+        Objects.requireNonNull(tag, "function tag must not be null");
+        Objects.requireNonNull(function, "function must not be null");
+        addFunction(tag, function, null);
+    }
+
+    /**
+     * Maps a function to the function tag, for use by the behavior.
+     * This method will not override any previous mapping added by {@link #func(FunctionTag,Runnable)}.
+     *
+     * @param behavior the owner
+     * @param tag the function tag
+     * @param function the function
+     */
+    void func(IBehavior behavior, FunctionTag tag, Runnable function) {
+        Objects.requireNonNull(behavior, "behavior must not be null");
+        Objects.requireNonNull(tag, "tag must not be null");
+        Objects.requireNonNull(function, "function must not be null");
+        addFunction(tag, function, behavior);
+    }
+    
+    /**
+     * Link a key binding to the specified function tag.
+     * This method will override a mapping set by the behavior.
+     *
+     * @param k the key binding
+     * @param tag the function tag
+     */
+    public void key(KeyBinding2 k, FunctionTag tag) {
```

Comment:
Similar to previous comment, can we change this to easily understandable name since it is a public method? Like `setKey` or `mapKeyBinding`?

Reply:
also, 
map() -> addHandler()
mapTail() -> addHandlerTail()
or something?

Reply:
Sure. I believe it will make it easy to understand the API functionality.